### PR TITLE
#141 - ethics v2

### PR DESCRIPTION
**Pull request by caravinci**
Completes the functionality for Ethics Letter

### CODE REVIEW

File: components/pages/Applications/ApplicationForm/Forms/validations/schemas.ts [hunk 1]
```diff
@@ -118,12 +118,21 @@ export const dataAccessAgreementSchema = yup.object().shape({
...
-    then: yup.array().min(1),
-    otherwise: yup.array().max(0),
-  }),
+  declaredAsRequired: yup.boolean().required(),
+  approvalLetterDocs: yup
+    .array(
+      yup.object().shape({
+        name: yup.string(),
+        objectId: yup.string(),
+        uploadedAtUtc: yup.string(),
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/schemas.ts:127]: First, what are we dealing with here at the Ethics section?
An array of objects, that relies on a boolean for the array to be required or not.

File: components/pages/Applications/ApplicationForm/Forms/EthicsLetter/index.tsx [hunk 2]
```diff
@@ -39,8 +71,7 @@ const EthicsLetter = ({
           className="vertical"
           disabled={isSectionDisabled}
           error={!!localState.declaredAsRequired?.error}
-          required={true}
-          // required={isRequired(localState.declaredAsRequired)}
+          required={isRequired(localState.declaredAsRequired)}
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/EthicsLetter/index.tsx:74]: so we make that boolean be required...

File: components/pages/Applications/ApplicationForm/Forms/EthicsLetter/index.tsx [hunk 3]
```diff
@@ -50,13 +81,12 @@ const EthicsLetter = ({
...
+            id="declaredAsRequired"
+            isChecked={isChecked}
+            onChange={handleSelectedRadioValueChange}
           >
-            <FormRadio disabled={isSectionDisabled} value="false" checked>
-              {FORM_TEXT.ethics.declarationOptions.notRequired}
-            </FormRadio>
-            <FormRadio disabled value="true">
+            <FormRadio value={false}>{FORM_TEXT.ethics.declarationOptions.notRequired}</FormRadio>
+            <FormRadio value={true}>
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/EthicsLetter/index.tsx:89]: Make its UI component work correctly

File: components/pages/Applications/ApplicationForm/Forms/EthicsLetter/index.tsx [hunk 4]
```diff
@@ -70,6 +100,15 @@ const EthicsLetter = ({
...
           <FormHelperText onErrorOnly>{localState.declaredAsRequired?.error?.[0]}</FormHelperText>
         </FormControl>
+
+        {selectedRadioValue && (
+          <UploadsTable
+            isSectionDisabled={isSectionDisabled}
+            localState={localState}
+            refetchAllData={refetchAllData}
+            required={isRequired(localState.approvalLetterDocs)}
+          />
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/EthicsLetter/index.tsx:110]: and allow uploading files, if the boolean is set to `true`.

File: components/pages/Applications/ApplicationForm/Forms/EthicsLetter/UploadsTable.tsx [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+} from '../types';
+
+const VALID_FILE_TYPE = [
+  'application/msword',
+  'application/pdf',
+  'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
+];
+const MAX_FILE_SIZE = 5242880;
+
+const UploadsTable = ({
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/EthicsLetter/UploadsTable.tsx:35]: If they want to upload files, we show them this pretty puppy

File: components/pages/Applications/ApplicationForm/Forms/EthicsLetter/UploadsTable.tsx [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+          type: 'updating',
+          value: data,
+        }),
+      )
+      .catch((err: AxiosError) => {
+        console.error('File could not be deleted.', err);
+      });
+  };
+
+  const handleFileUpload = (event: any) => {
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/EthicsLetter/UploadsTable.tsx:78]: this little piggy uploads files and triggers a global application state update with the response

File: components/pages/Applications/ApplicationForm/Forms/EthicsLetter/UploadsTable.tsx [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+  // make button work as input
+  const selectFile = () => {
+    const fp = fileInputRef.current;
+    if (fp) {
+      fp.click();
+    }
+  };
+
+  const handleFileDelete = (fileId: string) => (event: any) => {
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/EthicsLetter/UploadsTable.tsx:62]: while this one removes the files, updates the state as well

File: components/pages/Applications/ApplicationForm/Forms/Signature.tsx [hunk 8]
```diff
@@ -262,7 +261,7 @@ const Signature = ({ appId }: { appId: string }): ReactElement => {
                 Upload a file
               </Button>
 
-              <FormFieldHelpBubble text="Allowed file types: pdf. | Max file size: 200MB" />
+              <FormFieldHelpBubble text="Allowed file types: pdf. | Max file size: 5MB" />
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/Signature.tsx:264]: here, we update the new max size, as discussed at planning on July 15 2021

File: components/pages/Applications/ApplicationForm/Forms/Terms.tsx [hunk 9]
```diff
@@ -19,7 +16,7 @@ const Introduction = ({
   validateFieldTouched,
 }: {
   isSectionDisabled: boolean;
-  localState: FormSectionValidationState_Introduction;
+  localState: FormSectionValidationState_Terms;
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/Terms.tsx:19]: renamed for consistency

File: components/pages/Applications/ApplicationForm/Forms/helpers.tsx [hunk 10]
```diff
@@ -35,11 +36,12 @@ export const sectionSelector = ({
 }) => {
   const SectionComponent = sectionsData[selectedSection]?.component;
   const {
-    fields: storedFields = {},
+    fields: storedFields,
     meta: { overall },
-  } = formState.sections[selectedSection] || {};
+  }: FormSectionValidationState_SectionBase = formState.sections[selectedSection] || {};
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/helpers.tsx:41]: typing coverage

File: components/pages/Applications/ApplicationForm/Forms/helpers.tsx [hunk 11]
```diff
@@ -35,11 +36,12 @@ export const sectionSelector = ({
...
 
-  const isSectionDisabled = [FORM_STATES.DISABLED, FORM_STATES.LOCKED].includes(overall);
+  const isSectionDisabled =
+    !overall || [FORM_STATES.DISABLED, FORM_STATES.LOCKED].includes(overall);
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/helpers.tsx:44]: code guarding

File: components/pages/Applications/ApplicationForm/Forms/helpers.tsx [hunk 12]
```diff
@@ -55,6 +57,7 @@ export const sectionSelector = ({
     <SectionComponent
       isSectionDisabled={isSectionDisabled}
       localState={localState}
+      refetchAllData={formState.__refetchAllData}
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/helpers.tsx:60]: and giving the sections a way to trigger a state refetch from DB

File: components/pages/Applications/ApplicationForm/Forms/types.ts [hunk 13]
```diff
@@ -120,13 +120,9 @@ export type FormSectionValidationState_DataAccessAgreements =
   }>;
 export type FormSectionValidationState_EthicsLetter =
   FormSectionValidationState_SectionsGenericType<{
-    declaredAsRequired: { value: boolean };
+    declaredAsRequired: { value: boolean | null };
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/types.ts:123]: to match API

File: components/pages/Applications/ApplicationForm/Forms/types.ts [hunk 14]
```diff
@@ -173,18 +169,21 @@ export type FormSectionValidationState_Representative =
   }>;
 export type FormSectionValidationState_Signature =
   FormSectionValidationState_SectionsGenericType<{}>;
+export type FormSectionValidationState_Terms = FormSectionValidationState_SectionsGenericType<{
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/types.ts:172]: renamed for consistency

File: components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts [hunk 15]
```diff
@@ -44,6 +44,7 @@ export const getFieldDataFromEvent: FormFieldDataFromEvent = (event) => {
           value: event.target.checked,
         };
 
+      case 'radio': // this many need refactoring after improving UiKit Radio components
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:47]: handle a new field type 🎉

File: components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts [hunk 16]
```diff
@@ -89,21 +90,24 @@ export const schemaValidator = (fieldSchema: any, value: any) =>
 const getSeedValueByFieldType = (fieldType: string, fieldBase: any, seedValue: any) => {
   switch (fieldType) {
     case 'array': {
+      const isPublicationURLsArray = fieldBase.meta?.shape === 'publicationURLsArray';
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:93]: getting serious/specific with the validation

File: components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts [hunk 17]
```diff
@@ -142,21 +146,24 @@ export const getValueByFieldTypeToPublish = (
...
         }
       }
 
       return {
-        [fieldName]: fieldNameInner
-          ? { [fieldNameInner]: { accepted: value } }
-          : { accepted: value },
+        [fieldNameSource]: value,
       };
+    }
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:166]: so we can address the three types of booleans in this project: 🙄 
- acceptance
- arrays of acceptances
- plain JS boolean

File: components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts [hunk 18]
```diff
@@ -180,7 +187,7 @@ export const getValueByFieldTypeToPublish = (
     }
 
     case 'string':
-      return { [fieldName]: fieldNameInner ? { [fieldNameInner]: value } : value };
+      return { [fieldName]: fieldNameInner ? { [fieldNameInner]: value.trim() } : value.trim() };
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/helpers.ts:190]: bonus scope: trim strings

File: components/pages/Applications/ApplicationForm/Forms/validations/index.ts [hunk 19]
```diff
@@ -1,5 +1,6 @@
 import { Dispatch, useCallback, useEffect, useReducer, useState } from 'react';
-import { isEqual, merge, omit } from 'lodash';
+import { isEqual, omit } from 'lodash';
+import merge from 'deepmerge';
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/index.ts:3]: better merge. lodash's subpar

File: components/pages/Applications/ApplicationForm/Forms/validations/index.ts [hunk 20]
```diff
@@ -303,27 +304,32 @@ export const useFormValidation = (appId: string) => {
           {},
         ),
       },
+      __refetchAllData: (action: FormValidationAction) => {
+        action ? validationDispatch(action) : setTriggerFetch(true);
+      },
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/index.ts:309]: Now we can trigger a refetch on demand, or update the state based upon individual responses from API, outside my validation flow

File: components/pages/Applications/ApplicationForm/Forms/validations/index.ts [hunk 21]
```diff
@@ -358,22 +364,30 @@ export const useLocalValidation = (
         ...prev,
         [sectionName]: {
           ...prev[sectionName],
-          fields: merge(currentFields, storedFields),
+          fields: merge(currentFields, storedFields, {
+            customMerge: (key) => {
+              if (key === 'value') {
+                return (oldValue, newValue) => newValue;
+              }
+            },
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/index.ts:372]: merging objects to keep the validation properties, and existing validation errors when necessary; but replacing values as given by API.

File: components/pages/Applications/ApplicationForm/Forms/validations/index.ts [hunk 22]
```diff
@@ -358,22 +364,30 @@ export const useLocalValidation = (
...
-
       const [fieldName, fieldIndex, fieldOverride] = field.split('--');
       const currentSectionData = localState[sectionName];
       const currentSectionFields = currentSectionData?.fields;
       const currentField = currentSectionFields[fieldName];
       const oldValue = currentField.value;
 
+      fieldOverride === 'remove' ||
+        fieldsTouched.has(field) ||
+        setFieldTouched((prev) => new Set(prev.add(field)));
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/validations/index.ts:389]: making sure that removing a value doesn't leave buggy traces

**Review by caravinci — COMMENTED** [review]
A few comments for a large set of changes.

File: components/pages/Applications/ApplicationForm/Forms/EthicsLetter/UploadsTable.tsx [hunk 23]
```diff
@@ -0,0 +1,284 @@
...
+                width: 60,
+              },
+            ]}
+            css={css`
+              margin-top: 10px;
+            `}
+            data={localState.approvalLetterDocs?.value}
+            defaultSorted={[{ id: 'uploadedAtUtc', desc: true }]}
+            parentRef={containerRef}
+            showPagination={false}
```
Comment (caravinci) [inline components/pages/Applications/ApplicationForm/Forms/EthicsLetter/UploadsTable.tsx:272]: And lastly this little infuriating line! 
Why is that thing turned on by default!?

Feels like back to Bootstrap days overriding defaults!

File: components/pages/Applications/ApplicationForm/Forms/EthicsLetter/UploadsTable.tsx (hunk 6 above)
Comment (samrichca) [reply components/pages/Applications/ApplicationForm/Forms/EthicsLetter/UploadsTable.tsx:80]: 🐷

**Review by samrichca — APPROVED** [review]

**Pull request merged**